File: bindings/go/stores/spanner/spanner.go
package spanner

import (
	"bytes"
	"context"
	"errors"
	"sort"
	"strings"

	prolly "build.crab/prolly-go"
	gspanner "cloud.google.com/go/spanner"
	"google.golang.org/api/iterator"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
)

const (
	nodesTable = "ProllyNodes"
	hintsTable = "ProllyHints"
	rootsTable = "ProllyRoots"
)

type MutationKind uint8

const (
	MutationUpsertNode MutationKind = iota + 1
	MutationDeleteNode
	MutationUpsertHint
	MutationUpsertRoot
	MutationDeleteRoot
)

type Mutation struct {
	Kind      MutationKind
	Key       []byte
	Namespace []byte
	Value     []byte
}

type RootRecord struct {
	Name     []byte
	Manifest []byte
}

// Transaction is the narrow read/write transaction surface Store requires.
type Transaction interface {
	GetRoot(context.Context, []byte) (prolly.OptionalBytes, error)
	Buffer([]Mutation) error
}

// Client is the narrow provider surface wrapped around *spanner.Client.
type Client interface {
	GetNode(context.Context, []byte) (prolly.OptionalBytes, error)
	GetHint(context.Context, []byte, []byte) (prolly.OptionalBytes, error)
	GetRoot(context.Context, []byte) (prolly.OptionalBytes, error)
	ListNodeCIDs(context.Context) ([][]byte, error)
	ListRoots(context.Context) ([]RootRecord, error)
	Apply(context.Context, []Mutation) error
	ReadWrite(context.Context, func(context.Context, Transaction) error) error
}

type Options struct {
	AdapterName     string
	ReadParallelism uint32
}

type Store struct {
	client  Client
	options Options
}

func New(client *gspanner.Client, options Options) *Store {
	if client == nil {
		return NewWithClient(nil, options)
	}
	return NewWithClient(&sdkClient{client: client}, options)
}

func NewWithClient(client Client, options Options) *Store {
	if strings.TrimSpace(options.AdapterName) == "" {
		options.AdapterName = "spanner-v1"
	}
	if options.ReadParallelism == 0 {
		options.ReadParallelism = 16
	}
	return &Store{client: client, options: options}
}

func (s *Store) Descriptor(ctx context.Context) (prolly.StoreDescriptor, error) {
	if err := s.ready(ctx); err != nil {
		return prolly.StoreDescriptor{}, err
	}
	return prolly.StoreDescriptor{
		ProtocolMajor: prolly.StoreProtocolMajor, AdapterName: s.options.AdapterName, Provider: "spanner", SchemaVersion: 1,
		Capabilities: prolly.StoreCapabilities{
			NativeBatchReads: false, AtomicBatchWrites: true, NodeScan: true, Hints: true,
			AtomicNodesAndHint: true, RootScan: true, RootCompareAndSwap: true,
			Transactions: true, ReadParallelism: s.options.ReadParallelism,
		},
	}, nil
}

func (s *Store) GetNode(ctx context.Context, key []byte) (prolly.OptionalBytes, error) {
	if err := s.ready(ctx); err != nil {
		return prolly.OptionalBytes{}, err
	}
	value, err := s.client.GetNode(ctx, key)
	return value, spannerError("get_node", err)
}

func (s *Store) PutNode(ctx context.Context, key, value []byte) error {
	return s.apply(ctx, []Mutation{{Kind: MutationUpsertNode, Key: clone(key), Value: clone(value)}}, "put_node")
}

func (s *Store) DeleteNode(ctx context.Context, key []byte) error {
	return s.apply(ctx, []Mutation{{Kind: MutationDeleteNode, Key: clone(key)}}, "delete_node")
}

func (s *Store) BatchNodes(ctx context.Context, mutations []prolly.NodeMutation) error {
	providerMutations := make([]Mutation, len(mutations))
	for index, mutation := range mutations {
		if mutation.Value.Present {
			providerMutations[index] = Mutation{Kind: MutationUpsertNode, Key: clone(mutation.Key), Value: clone(mutation.Value.Value)}
		} else {
			providerMutations[index] = Mutation{Kind: MutationDeleteNode, Key: clone(mutation.Key)}
		}
	}
	return s.apply(ctx, providerMutations, "batch_nodes")
}

func (s *Store) PublishNodes(ctx context.Context, publication prolly.NodePublication) error {
	return prolly.PublishNodesWithGeneralPath(ctx, s, publication)
}

func (s *Store) BatchGetNodesOrdered(ctx context.Context, keys [][]byte) ([]prolly.OptionalBytes, error) {
	result := make([]prolly.OptionalBytes, len(keys))
	for index, key := range keys {
		value, err := s.GetNode(ctx, key)
		if err != nil {
			return nil, err
		}
		result[index] = value
	}
	return result, nil
}

func (s *Store) ListNodeCIDs(ctx context.Context) ([][]byte, error) {
	if err := s.ready(ctx); err != nil {
		return nil, err
	}
	keys, err := s.client.ListNodeCIDs(ctx)
	if err != nil {
		return nil, spannerError("list_nodes", err)
	}
	result := make([][]byte, 0, len(keys))
	for _, key := range keys {
		if len(key) == 32 {
			result = append(result, clone(key))
		}
	}
	sort.Slice(result, func(i, j int) bool { return bytes.Compare(result[i], result[j]) < 0 })
	return result, nil
}

func (s *Store) GetHint(ctx context.Context, namespace, key []byte) (prolly.OptionalBytes, error) {
	if err := s.ready(ctx); err != nil {
		return prolly.OptionalBytes{}, err
	}
	value, err := s.client.GetHint(ctx, namespace, key)
	return value, spannerError("get_hint", err)
}

func (s *Store) PutHint(ctx context.Context, namespace, key, value []byte) error {
	return s.apply(ctx, []Mutation{{Kind: MutationUpsertHint, Namespace: clone(namespace), Key: clone(key), Value: clone(value)}}, "put_hint")
}

func (s *Store) BatchPutNodesWithHint(ctx context.Context, nodes []prolly.NodeEntry, namespace, key, value []byte) error {
	mutations := make([]Mutation, 0, len(nodes)+1)
	for _, node := range nodes {
		mutations = append(mutations, Mutation{Kind: MutationUpsertNode, Key: clone(node.Key), Value: clone(node.Value)})
	}
	mutations = append(mutations, Mutation{Kind: MutationUpsertHint, Namespace: clone(namespace), Key: clone(key), Value: clone(value)})
	return s.apply(ctx, mutations, "batch_nodes_hint")
}

func (s *Store) GetRootManifest(ctx context.Context, name []byte) (prolly.OptionalBytes, error) {
	if err := s.ready(ctx); err != nil {
		return prolly.OptionalBytes{}, err
	}
	value, err := s.client.GetRoot(ctx, name)
	return value, spannerError("get_root", err)
}

func (s *Store) PutRootManifest(ctx context.Context, name, manifest []byte) error {
	return s.apply(ctx, []Mutation{{Kind: MutationUpsertRoot, Key: clone(name), Value: clone(manifest)}}, "put_root")
}

func (s *Store) DeleteRootManifest(ctx context.Context, name []byte) error {
	return s.apply(ctx, []Mutation{{Kind: MutationDeleteRoot, Key: clone(name)}}, "delete_root")
}

func (s *Store) CompareAndSwapRootManifest(ctx context.Context, name []byte, expected, replacement prolly.OptionalBytes) (result prolly.RootCASResult, err error) {
	if err := s.ready(ctx); err != nil {
		return result, err
	}
	err = s.client.ReadWrite(ctx, func(txCtx context.Context, tx Transaction) error {
		current, readErr := tx.GetRoot(txCtx, name)
		if readErr != nil {
			return readErr
		}
		if !optionalEqual(current, expected) {
			result = prolly.RootCASResult{Current: current}
			return nil
		}
		mutation := Mutation{Kind: MutationDeleteRoot, Key: clone(name)}
		if replacement.Present {
			mutation.Kind, mutation.Value = MutationUpsertRoot, clone(replacement.Value)
		}
		if bufferErr := tx.Buffer([]Mutation{mutation}); bufferErr != nil {
			return bufferErr
		}
		result = prolly.RootCASResult{Applied: true, Current: replacement.Clone()}
		return nil
	})
	return result, spannerError("root_cas", err)
}

func (s *Store) ListRootManifests(ctx context.Context) ([]prolly.NamedStoreRoot, error) {
	if err := s.ready(ctx); err != nil {
		return nil, err
	}
	records, err := s.client.ListRoots(ctx)
	if err != nil {
		return nil, spannerError("list_roots", err)
	}
	result := make([]prolly.NamedStoreRoot, len(records))
	for index, record := range records {
		result[index] = prolly.NamedStoreRoot{Name: clone(record.Name), Manifest: clone(record.Manifest)}
	}
	sort.Slice(result, func(i, j int) bool { return bytes.Compare(result[i].Name, result[j].Name) < 0 })
	return result, nil
}

func (s *Store) CommitTransaction(ctx context.Context, nodes []prolly.NodeMutation, conditions []prolly.RootCondition, roots []prolly.RootWrite) (result prolly.StoreTransactionResult, err error) {
	if err := s.ready(ctx); err != nil {
		return result, err
	}
	err = s.client.ReadWrite(ctx, func(txCtx context.Context, tx Transaction) error {
		for _, condition := range conditions {
			current, readErr := tx.GetRoot(txCtx, condition.Name)
			if readErr != nil {
				return readErr
			}
			if !optionalEqual(current, condition.Expected) {
				result = prolly.StoreTransactionResult{Conflict: &prolly.StoreTransactionConflict{Name: clone(condition.Name), Expected: condition.Expected.Clone(), Current: current}}
				return nil
			}
		}
		mutations := make([]Mutation, 0, len(nodes)+len(roots))
		for _, node := range nodes {
			mutation := Mutation{Kind: MutationDeleteNode, Key: clone(node.Key)}
			if node.Value.Present {
				mutation.Kind, mutation.Value = MutationUpsertNode, clone(node.Value.Value)
			}
			mutations = append(mutations, mutation)
		}
		for _, root := range roots {
			mutation := Mutation{Kind: MutationDeleteRoot, Key: clone(root.Name)}
			if root.Replacement.Present {
				mutation.Kind, mutation.Value = MutationUpsertRoot, clone(root.Replacement.Value)
			}
			mutations = append(mutations, mutation)
		}
		if len(mutations) != 0 {
			if err := tx.Buffer(mutations); err != nil {
				return err
			}
		}
		result = prolly.StoreTransactionResult{Applied: true}
		return nil
	})
	return result, spannerError("transaction", err)
}

func (s *Store) apply(ctx context.Context, mutations []Mutation, operation string) error {
	if err := s.ready(ctx); err != nil {
		return err
	}
	if len(mutations) == 0 {
		return nil
	}
	return spannerError(operation, s.client.Apply(ctx, mutations))
}

func (s *Store) ready(ctx context.Context) error {
	if err := ctx.Err(); err != nil {
		return err
	}
	if s == nil || s.client == nil {
		return &prolly.StoreError{Code: "invalid_configuration", Message: "Spanner client is nil"}
	}
	return nil
}

type sdkClient struct{ client *gspanner.Client }

func (c *sdkClient) GetNode(ctx context.Context, key []byte) (prolly.OptionalBytes, error) {
	return readSDKValue(ctx, c.client.Single(), nodesTable, gspanner.Key{clone(key)}, "Node")
}
func (c *sdkClient) GetHint(ctx context.Context, namespace, key []byte) (prolly.OptionalBytes, error) {
	return readSDKValue(ctx, c.client.Single(), hintsTable, gspanner.Key{clone(namespace), clone(key)}, "Value")
}
func (c *sdkClient) GetRoot(ctx context.Context, name []byte) (prolly.OptionalBytes, error) {
	return readSDKValue(ctx, c.client.Single(), rootsTable, gspanner.Key{clone(name)}, "Manifest")
}
func (c *sdkClient) ListNodeCIDs(ctx context.Context) ([][]byte, error) {
	rows := c.client.Single().Query(ctx, gspanner.Statement{SQL: "SELECT Cid FROM ProllyNodes ORDER BY Cid"})
	defer rows.Stop()
	var result [][]byte
	for {
		row, err := rows.Next()
		if err == iterator.Done {
			return result, nil
		}
		if err != nil {
			return nil, err
		}
		var key []byte
		if err := row.ColumnByName("Cid", &key); err != nil {
			return nil, err
		}
		result = append(result, clone(key))
	}
}
func (c *sdkClient) ListRoots(ctx context.Context) ([]RootRecord, error) {
	rows := c.client.Single().Query(ctx, gspanner.Statement{SQL: "SELECT Name, Manifest FROM ProllyRoots ORDER BY Name"})
	defer rows.Stop()
	var result []RootRecord
	for {
		row, err := rows.Next()
		if err == iterator.Done {
			return result, nil
		}
		if err != nil {
			return nil, err
		}
		var name, manifest []byte
		if err := row.Columns(&name, &manifest); err != nil {
			return nil, err
		}
		result = append(result, RootRecord{Name: clone(name), Manifest: clone(manifest)})
	}
}
func (c *sdkClient) Apply(ctx context.Context, mutations []Mutation) error {
	_, err := c.client.Apply(ctx, sdkMutations(mutations))
	return err
}
func (c *sdkClient) ReadWrite(ctx context.Context, function func(context.Context, Transaction) error) error {
	_, err := c.client.ReadWriteTransaction(ctx, func(txCtx context.Context, tx *gspanner.ReadWriteTransaction) error {
		return function(txCtx, &sdkTransaction{tx: tx})
	})
	return err
}

type sdkTransaction struct {
	tx *gspanner.ReadWriteTransaction
}

func (t *sdkTransaction) GetRoot(ctx context.Context, name []byte) (prolly.OptionalBytes, error) {
	return readSDKValue(ctx, t.tx, rootsTable, gspanner.Key{clone(name)}, "Manifest")
}
func (t *sdkTransaction) Buffer(mutations []Mutation) error {
	return t.tx.BufferWrite(sdkMutations(mutations))
}

type rowReader interface {
	ReadRow(context.Context, string, gspanner.Key, []string) (*gspanner.Row, error)
}

func readSDKValue(ctx context.Context, reader rowReader, table string, key gspanner.Key, column string) (prolly.OptionalBytes, error) {
	row, err := reader.ReadRow(ctx, table, key, []string{column})
	if errors.Is(err, gspanner.ErrRowNotFound) {
		return prolly.MissingBytes(), nil
	}
	if err != nil {
		return prolly.OptionalBytes{}, err
	}
	var value []byte
	if err := row.Column(0, &value); err != nil {
		return prolly.OptionalBytes{}, err
	}
	return prolly.PresentBytes(value), nil
}

func sdkMutations(mutations []Mutation) []*gspanner.Mutation {
	result := make([]*gspanner.Mutation, 0, len(mutations))
	for _, mutation := range mutations {
		switch mutation.Kind {
		case MutationUpsertNode:
			result = append(result, gspanner.InsertOrUpdate(nodesTable, []string{"Cid", "Node"}, []any{clone(mutation.Key), clone(mutation.Value)}))
		case MutationDeleteNode:
			result = append(result, gspanner.Delete(nodesTable, gspanner.Key{clone(mutation.Key)}))
		case MutationUpsertHint:
			result = append(result, gspanner.InsertOrUpdate(hintsTable, []string{"Namespace", "HintKey", "Value"}, []any{clone(mutation.Namespace), clone(mutation.Key), clone(mutation.Value)}))
		case MutationUpsertRoot:
			result = append(result, gspanner.InsertOrUpdate(rootsTable, []string{"Name", "Manifest"}, []any{clone(mutation.Key), clone(mutation.Value)}))
		case MutationDeleteRoot:
			result = append(result, gspanner.Delete(rootsTable, gspanner.Key{clone(mutation.Key)}))
		}
	}
	return result
}

func spannerError(operation string, err error) error {
	if err == nil {
		return nil
	}
	if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
		return err
	}
	code := status.Code(err)
	retryable := code == codes.Aborted || code == codes.Unavailable || code == codes.ResourceExhausted || code == codes.DeadlineExceeded
	return &prolly.StoreError{Code: "provider_error", Message: operation + ": " + err.Error(), Retryable: retryable, ProviderCode: code.String(), Cause: err}
}

func optionalEqual(left, right prolly.OptionalBytes) bool {
	return left.Present == right.Present && (!left.Present || bytes.Equal(left.Value, right.Value))
}
func clone(value []byte) []byte {
	if value == nil {
		return nil
	}
	result := make([]byte, len(value))
	copy(result, value)
	return result
}

File: bindings/ruby/stores/spanner/lib/prolly/store/spanner.rb
# frozen_string_literal: true

require 'google/cloud/spanner'
require 'prolly'
require 'stringio'

module Prolly
  class SpannerRemoteStore < ForeignRemoteStore
    DDL = [
      "CREATE TABLE ProllyNodes (\n  Cid BYTES(32) NOT NULL,\n  Node BYTES(MAX) NOT NULL\n) PRIMARY KEY (Cid)",
      "CREATE TABLE ProllyHints (\n  Namespace BYTES(MAX) NOT NULL,\n  HintKey BYTES(MAX) NOT NULL,\n  Value BYTES(MAX) NOT NULL\n) PRIMARY KEY (Namespace, HintKey)",
      "CREATE TABLE ProllyRoots (\n  Name BYTES(MAX) NOT NULL,\n  Manifest BYTES(MAX) NOT NULL\n) PRIMARY KEY (Name)"
    ].freeze

    def initialize(client, narrow_client: false)
      unless narrow_client || client.is_a?(Google::Cloud::Spanner::Client)
        raise ArgumentError, 'client must be Google::Cloud::Spanner::Client'
      end

      @client = narrow_client ? client : SdkClient.new(client)
      @closed = false
    end

    def self.from_client(client)
      raise ArgumentError, 'Cloud Spanner client is required' unless client
      new(client, narrow_client: true)
    end

    def close = @closed = true

    def descriptor
      value = StoreDescriptorRecord.new(
        protocol_major: STORE_PROTOCOL_MAJOR, adapter_name: 'spanner-v1', provider: 'spanner', schema_version: 1,
        capabilities: StoreCapabilitiesRecord.new(
          native_batch_reads: false, atomic_batch_writes: true, node_scan: true,
          hints: true, atomic_nodes_and_hint: true, root_scan: true,
          root_compare_and_swap: true, transactions: true, read_parallelism: 1
        ),
        limits: StoreLimitsRecord.new(
          max_batch_read_items: nil, max_batch_write_items: nil,
          max_transaction_operations: nil, max_node_bytes: nil
        )
      )
      StoreDescriptorResultRecord.new(value: value, error: nil)
    end

    def get_node(cid) = optional_call { @client.get_node(cid.b) }
    def put_node(cid, value) = unit_call { apply([[:upsert_node, cid.b, value.b]]) }
    def delete_node(cid) = unit_call { apply([[:delete_node, cid.b]]) }

    def batch_nodes(operations)
      mutations = operations.map do |item|
        item.value.present ? [:upsert_node, item.key.b, item.value.value.b] : [:delete_node, item.key.b]
      end
      unit_call { apply(mutations) }
    end

    def batch_get_nodes_ordered(cids)
      values = ensure_open { cids.map { |cid| optional(@client.get_node(cid.b)) } }
      OptionalBytesListResultRecord.new(values: values, error: nil)
    rescue StandardError => error
      OptionalBytesListResultRecord.new(values: [], error: store_error(error))
    end

    def list_node_cids
      values = ensure_open { @client.list_node_cids.filter { |cid| cid.bytesize == 32 }.sort }
      BytesListResultRecord.new(values: values, error: nil)
    rescue StandardError => error
      BytesListResultRecord.new(values: [], error: store_error(error))
    end

    def get_hint(namespace, hint_key) = optional_call { @client.get_hint(namespace.b, hint_key.b) }
    def put_hint(namespace, hint_key, value) = unit_call { apply([[:upsert_hint, namespace.b, hint_key.b, value.b]]) }

    def batch_put_nodes_with_hint(nodes, namespace, hint_key, value)
      mutations = nodes.map { |node| [:upsert_node, node.key.b, node.value.b] }
      mutations << [:upsert_hint, namespace.b, hint_key.b, value.b]
      unit_call { apply(mutations) }
    end

    def get_root_manifest(name) = optional_call { @client.get_root(name.b) }
    def put_root_manifest(name, manifest) = unit_call { apply([[:upsert_root, name.b, manifest.b]]) }
    def delete_root_manifest(name) = unit_call { apply([[:delete_root, name.b]]) }

    def compare_and_swap_root_manifest(name, expected, replacement)
      result = ensure_open do
        @client.read_write do |transaction|
          current = transaction.get_root(name.b)
          unless matches?(current, expected)
            next RootCasResultRecord.new(applied: false, current: optional(current), error: nil)
          end
          transaction.buffer([root_mutation(name.b, replacement)])
          RootCasResultRecord.new(applied: true, current: replacement, error: nil)
        end
      end
      result
    rescue StandardError => error
      RootCasResultRecord.new(applied: false, current: optional(nil), error: store_error(error))
    end

    def list_root_manifests
      values = ensure_open do
        @client.list_roots.sort_by(&:first).map do |name, manifest|
          NamedBytesRecord.new(name: name, value: manifest)
        end
      end
      NamedBytesListResultRecord.new(values: values, error: nil)
    rescue StandardError => error
      NamedBytesListResultRecord.new(values: [], error: store_error(error))
    end

    def commit_transaction(nodes, conditions, roots)
      result = ensure_open do
        @client.read_write do |transaction|
          conflict = conditions.filter_map do |condition|
            current = transaction.get_root(condition.name.b)
            next if matches?(current, condition.expected)
            StoreTransactionConflictRecord.new(
              name: condition.name.b, expected: condition.expected, current: optional(current)
            )
          end.first
          next TransactionResultRecord.new(applied: false, conflict: conflict, error: nil) if conflict

          mutations = nodes.map do |node|
            node.value.present ? [:upsert_node, node.key.b, node.value.value.b] : [:delete_node, node.key.b]
          end
          roots.each { |root| mutations << root_mutation(root.name.b, root.replacement) }
          transaction.buffer(mutations)
          TransactionResultRecord.new(applied: true, conflict: nil, error: nil)
        end
      end
      result
    rescue StandardError => error
      TransactionResultRecord.new(applied: false, conflict: nil, error: store_error(error))
    end

    private

    def apply(mutations) = ensure_open { @client.apply(mutations) }

    def ensure_open
      raise 'Cloud Spanner store is closed' if @closed
      yield
    end

    def optional(value) = OptionalBytesRecord.new(present: !value.nil?, value: value || ''.b)
    def matches?(current, expected) = expected.present ? current == expected.value : current.nil?
    def root_mutation(name, replacement) = replacement.present ? [:upsert_root, name, replacement.value.b] : [:delete_root, name]

    def optional_call
      OptionalBytesResultRecord.new(value: optional(ensure_open { yield }), error: nil)
    rescue StandardError => error
      OptionalBytesResultRecord.new(value: optional(nil), error: store_error(error))
    end

    def unit_call
      yield
      UnitResultRecord.new(error: nil)
    rescue StandardError => error
      UnitResultRecord.new(error: store_error(error))
    end

    def store_error(error)
      code = grpc_code(error)
      classification, retryable = case code
                                  when 8 then ['resource_exhausted', true]
                                  when 4, 10, 14 then ['unavailable', true]
                                  else ['internal', false]
                                  end
      StoreErrorRecord.new(
        code: classification, message: 'Cloud Spanner provider operation failed',
        retryable: retryable, provider_code: code ? "grpc:#{code}" : error.class.name
      )
    end

    def grpc_code(error)
      value = error.respond_to?(:code) ? error.code : nil
      return value if value.is_a?(Integer)
      return value.to_i if value.respond_to?(:to_i) && value.to_i.positive?

      names = { deadline_exceeded: 4, resource_exhausted: 8, aborted: 10, unavailable: 14 }
      names[value.to_s.downcase.to_sym]
    end

    class SdkClient
      def initialize(client) = @client = client

      def get_node(key) = read_value(@client, 'ProllyNodes', [:Node], [key])
      def get_hint(namespace, key) = read_value(@client, 'ProllyHints', [:Value], [namespace, key])
      def get_root(name) = read_value(@client, 'ProllyRoots', [:Manifest], [name])

      def list_node_cids
        @client.read('ProllyNodes', [:Cid], keys: []).rows.map { |row| binary(row[0]) }
      end

      def list_roots
        @client.read('ProllyRoots', %i[Name Manifest], keys: []).rows.map do |row|
          [binary(row[0]), binary(row[1])]
        end
      end

      def apply(mutations)
        @client.commit { |commit| buffer_sdk(commit, mutations) }
      end

      def read_write
        result = nil
        @client.transaction { |transaction| result = yield SdkTransaction.new(transaction) }
        result
      end

      private

      def read_value(reader, table, columns, key)
        row = reader.read(table, columns, keys: [key.map { |value| io(value) }]).rows.first
        row ? binary(row[0]) : nil
      end

      def buffer_sdk(target, mutations)
        mutations.each do |mutation|
          case mutation[0]
          when :upsert_node
            target.upsert 'ProllyNodes', { Cid: io(mutation[1]), Node: io(mutation[2]) }
          when :delete_node
            target.delete 'ProllyNodes', [[io(mutation[1])]]
          when :upsert_hint
            target.upsert 'ProllyHints', { Namespace: io(mutation[1]), HintKey: io(mutation[2]), Value: io(mutation[3]) }
          when :upsert_root
            target.upsert 'ProllyRoots', { Name: io(mutation[1]), Manifest: io(mutation[2]) }
          when :delete_root
            target.delete 'ProllyRoots', [[io(mutation[1])]]
          else
            raise ArgumentError, 'unknown Cloud Spanner mutation'
          end
        end
      end

      def io(value) = StringIO.new(value.b)
      def binary(value) = value.respond_to?(:string) ? value.string.b : value.b

      class SdkTransaction
        def initialize(transaction) = @transaction = transaction

        def get_root(name)
          row = @transaction.read('ProllyRoots', [:Manifest], keys: [[StringIO.new(name.b)]]).rows.first
          value = row&.[](0)
          value.respond_to?(:string) ? value.string.b : value&.b
        end

        def buffer(mutations)
          SdkClient.allocate.send(:buffer_sdk, @transaction, mutations)
        end
      end
    end
  end
end
